plutus-core/docs/equalsCanonical.md

```diff
@@ -0,0 +1,203 @@
+        VConstr _ _ -> EqCanonicalFalse
+        _           -> EqCanonicalUnexpectedInput
+    eqCanonical _ = \_ -> EqCanonicalUnexpectedInput
+```
+
+I.e. for `CekValue` `eqCanonical` checks if both the arguments are either `VConstr` or `VCon` and fails otherwise. If the constructors are the same, then the function checks if their contents are the same too.
+
+An `eqCanonical` application can have weird semantics if it doesn't come from a well-typed term, for example
+
+```haskell
+eqCanonical
+    (VConstr 0 (ConsStack (mkConstant () True) EmptyStack))
+    (VConstr 1 (ConsStack (VDelay <...> <...>) EmptyStack))
+```
+
+returns `False`, because the two tags are different (`0` vs `1`) and since `EqCanonicalResult` is short-circuiting we never get to check if the two `ArgStack`s are equal and thus never reach `VDelay`, which would otherwise fail evaluation.
+
+But that's the usual untyped evaluation stuff, we have plenty of tests that fail to type check but evaluate successfully, see `IllTypedRuns` tests in [this](https://github.com/input-output-hk/plutus/blob/173dce5ee85cb8038563dd39299abb550ea13b88/plutus-core/untyped-plutus-core/test/Evaluation/Golden.hs) module. Preventing things from being weird is the reason why we have a type checker, hence let's improve upon the untyped builtin described in this section.
+
+## Typed
+
+An obvious starting point is to assign `equalsCanonical` the following type signature:
+
+```
+equalsCanonical : all a. a -> a -> bool
+```
+
+but that would be a parametricity breaking lie: the type signature says that the builtin can compare two values of an arbitrary type for equality and return a conclusive `True` or `False`, but this isn't of course possible if the values are, say, functions.
+
+Instead, we can introduce a new built-in type constructor, let's call it `canonical` with the semantics that every `canonical a` value is known to be in canonical form. `equalsCanonical` then becomes
+
+```
+equalsCanonical : all a. canonical a -> canonical a -> bool
+```
+
+which is no longer a lie.
+
+On the Haskell side it'll look like this:
+
+```haskell
+    toBuiltinMeaning _ver EqualsCanonical =
+        makeBuiltinMeaning
+            equalsCanonicalPlc
+            (\_ _ _ -> ExBudgetLast mempty)
+      where
+        equalsCanonicalPlc :: Opaque val (Canonical a) -> Opaque val (Canonical a) -> EvaluationResult Bool
+        equalsCanonicalPlc val1 val2 = fromEqCanonicalResult $ eqCanonical val1 val2
+```
+
+Now we of course didn't resolve any problem by introducing `canonical`, we just moved it from `equalsCanonical` to whatever builtin is going to create `canonical` values. However moving the problem this way opens up some interesting opportunities. A reasonable way of constructing a `canonical` value is a builtin that takes some kind of a proof that the value can indeed be represented in canonical form, here's one option on how to achieve that:
+
+```
+mkCanonical : all a. (a -> data) -> a -> canonical a
```

Comment:
This function does not seem to provide any reassurance. Construct any `(d : data)`, then 
`mkCanonical (const d) x` constructs a `canonical a` for any `x : a`.

Reply:
> This function does not seem to provide any reassurance. Construct any `(d : data)`, then
`mkCanonical (const d) x` constructs a `canonical a` for any `x : a`.

I can construct `unsafeCoerce : all a b. a -> b`, I can even do it without using `error` (and without any extra builtins of course). Does it mean that our type system doesn't provide any reassurance?

Correct typing does not guarantee correct semantics and not every rule needs to be reflected at the type level. Those that aren't reflected at the type level are called "invariants" and "laws" and are very common in functional programming, including our own language. For a random example, users of `verifyEcdsaSecp256k1Signature` need to understand that they have to compute the hash themselves:

```
-- While this primitive /accepts/ a hash, any caller should only pass it hashes
-- that they computed themselves: specifically, they should receive the
-- /message/ from a sender and hash it, rather than receiving the /hash/ from
-- said sender. Failure to do so can be
-- [dangerous](https://bitcoin.stackexchange.com/a/81116/35586).
```

Good news are that users of `mkCanonical` are compilers authors, not script writers. The compiler authors need to make sure that the user cannot easily forge a bogus `a -> data` function and feed it to the builtin, which is stupidly easy to achieve in our case: don't expose the builtin directly and instead expose 

```haskell
mkCanonical :: ToData a => a -> Canonical a
```

in Plutus Tx. Given that `ToData` is derived by our tooling, it's about as safe as it gets when it comes to developing smart contracts.

I can assure you that smart contract developers don't worry about us not saving them enough from themselves, they worry about us not providing them with ways to write reasonable code that they need to write to achieve their goals. I'm not saying we should implement every request, but all risk is relative and in this case it's negligible compared to basically everything else that compiler and smart contract developers do _daily_.

And if we don't trust in ourselves and other compiler developers enough to be confident about being able to feed the right argument to a function in a single place, we should go into the woods are reunite with the nature or something instead.

Reply:
> I can construct `unsafeCoerce : all a b. a -> b`, I can even do it without using `error` (and without any extra builtins of course). Does it mean that our type system doesn't provide any reassurance?

BTW, How can you construct this from between the language?

Reply:
> The question is what is the point of the function `(a -> data)`?
> It's not evidence of anything.
> Why not make the type `mkCanonical : all a. a -> canonical a` directly?

These are reasonable and important questions, I'll elaborate on them in the doc itself.

> BTW, How can you construct this from within the language?

`fix id`. You get to review the [PR](https://github.com/input-output-hk/plutus/pull/5450) adding that example and the type dumping functionality (don't actually go through hundreds of golden files, only review the changes in `.hs` files). The full `unsafeCoerce` term is [here](https://github.com/input-output-hk/plutus/blob/92d1d2819afbfe797716f5ec4ddf664bea9a700e/plutus-core/plutus-core/test/Pretty/Golden/Readable/Examples/Function/unsafeCoerce.golden) and its type is the expected [one](https://github.com/input-output-hk/plutus/blob/92d1d2819afbfe797716f5ec4ddf664bea9a700e/plutus-core/plutus-core/test/TypeSynthesis/Golden/Examples/Function/unsafeCoerce.plc.golden).

Reply:
The problem that I have with `mkCanonical` is that `canonical` itself was introduced to avoid having a function
`equalsCanonical : all a. a -> a -> bool`, but that function is definable from `mkCanonical` with the expected parametricity-breaking semantics.

Reply:
I tend to agree with Mauro's argument here.

Reply:
@mjaskelioff 

> The problem that I have with `mkCanonical` is that canonical itself was introduced to avoid having a function
`equalsCanonical : all a. a -> a -> bool`

it wasn't. The original and the current text say

> Now we of course didn't resolve any problem by introducing `canonical`, we just moved it from `equalsCanonical` to whatever builtin is going to create `canonical` values. However moving the problem this way opens up some interesting opportunities.